tsdb/wal/wal.go

```diff
@@ -757,23 +769,52 @@ func listSegments(dir string) (refs []segmentRef, err error) {
 		return nil, err
 	}
 	var last int
+	// assume segments in sequence
+	inSequence := true
 	for _, fn := range files {
 		k, err := strconv.Atoi(fn)
 		if err != nil {
 			continue
 		}
-		if len(refs) > 0 && k > last+1 {
-			return nil, errors.New("segments are not sequential")
+
+		if len(refs) > 0 && k > last+1 && inSequence {
+			inSequence = false
 		}
 		refs = append(refs, segmentRef{name: fn, index: k})
 		last = k
 	}
+	if !inSequence {
+		return refs, errors.New(ErrorUnsequentialSegments)
+	}
 	sort.Slice(refs, func(i, j int) bool {
 		return refs[i].index < refs[j].index
 	})
 	return refs, nil
 }
 
+// Returns the sequential segments after deleting the out of sequence wals after the sequence gap.
+func (w *WAL) repairUnsequentialSegments() error {
+	segments, _ := listSegments(w.dir)
+	level.Warn(w.logger).Log("msg", "deleting unsequential segments")
```

Comment:
I would also add the last good segment index and the segments that would be deleted. which means moving this log a bit further down.

Reply:
sure